devito/passes/clusters/blocking.py

```diff
@@ -44,62 +46,206 @@ def blocking(clusters, options):
-        processed = Blocking(options).process(processed)
+        clusters = SynthesizeBlocking(sregistry, options).process(clusters)
 
     if options['skewing']:
-        processed = Skewing(options).process(processed)
+        clusters = SynthesizeSkewing(options).process(clusters)
 
-    return processed
+    return clusters
 
 
-class Blocking(Queue):
+class AnayzeBlockingBase(Queue):
 
-    template = "%s%d_blk%s"
+    """
+    Encode the TILABLE property.
+    """
+
+    def process(self, clusters):
+        return self._process_fatd(clusters, 1)
+
+    def _process_fatd(self, clusters, level, prefix=None):
+        # Truncate recursion in case of TILABLE, non-perfect sub-nests, as
+        # it's an unsupported case
+        if prefix:
+            d = prefix[-1].dim
+
+            if any(TILABLE in c.properties[d] for c in clusters) and \
+               len({c.itintervals[:level] for c in clusters}) > 1:
+                return clusters
+
+        return super()._process_fatd(clusters, level, prefix)
+
+
+class AnalyzeBlocking(AnayzeBlockingBase):
+
+    def callback(self, clusters, prefix):
+        if not prefix:
+            return clusters
+
+        d = prefix[-1].dim
+
+        for c in clusters:
+            if not {PARALLEL,
+                    PARALLEL_IF_ATOMIC,
+                    PARALLEL_IF_PVT}.intersection(c.properties[d]):
+                return clusters
+
+        # All good, `d` is actually TILABLE
+        processed = attach_property(clusters, d, TILABLE)
+
+        return processed
+
+
+class AnalyzeHeuristicBlocking(AnayzeBlockingBase):
 
     def __init__(self, options):
-        self.inner = bool(options['blockinner'])
+        super().__init__()
+
+        self.inner = options['blockinner']
+
+    def process(self, clusters):
+        clusters = super().process(clusters)
+
+        # Heuristic: if there aren't at least two TILABLE Dimensions, drop it
+        processed = []
+        for c in clusters:
+            ntilable = len([TILABLE for v in c.properties.values() if TILABLE in v])
+            if ntilable > 1:
+                processed.append(c)
+            else:
+                properties = {d: v - {TILABLE} for d, v in c.properties.items()}
+                processed.append(c.rebuild(properties=properties))
+
+        return processed
+
+    def callback(self, clusters, prefix):
+        if not prefix:
+            return clusters
+
+        d = prefix[-1].dim
+
+        for c in clusters:
+            # PARALLEL* and AFFINE are necessaary conditions
+            if AFFINE not in c.properties[d] or \
+               not ({PARALLEL, PARALLEL_IF_PVT} & c.properties[d]):
+                return clusters
+
+            # Heuristic: innermost Dimensions may be ruled out a-priori
+            is_inner = d is c.itintervals[-1].dim
+            if is_inner and not self.inner:
+                return clusters
+
+            # Heuristic: TILABLE not worth it if not within SEQUENTIAL Dimension
```

Comment:
# Heuristic: TILABLE is not worth it if not within a SEQUENTIAL Dimension

Reply:
fixing